alpa/device_mesh.py

```diff
@@ -146,12 +148,27 @@ def __init__(self, server_address: str, num_hosts: int, host_id: int,
             print("Use signal send recv.")
             self.signal_tensors = []
             for d in self.local_devices:
+                jax_tensor = device_put(jnp.ones((1,), dtype=jnp.int8), d)
                 self.signal_tensors.append(
-                    jax_tensor_to_cupy(device_put(
-                        jnp.ones((1,), dtype=jnp.int8), d),
-                                       take_ownership=True))
+                    jax_tensor_to_xla_buffer(jax_tensor) if global_config.
+                    nccl_mode == "from_xla_extension" else jax_tensor_to_cupy(
+                        jax_tensor, take_ownership=True))
+
         self.launched = True
 
+        if global_config.nccl_mode == "from_xla_extension":
+            self.send_tile_func = self.xla_nccl_send_tile
+            self.recv_tile_func = self.xla_nccl_recv_tile
+            self.allgather_func = self.xla_nccl_allgather
+            self.broadcast_func = self.xla_nccl_broadcast
+            self.nccl_local_allgather_init_comms = xe.nccl_init_communicator
+        else:
+            self.send_tile_func = self.cupy_nccl_send_tile
+            self.recv_tile_func = self.cupy_nccl_recv_tile
+            self.allgather_func = self.cupy_nccl_allgather
+            self.broadcast_func = self.cupy_nccl_broadcast
+            self.nccl_local_allgather_init_comms = nccl.NcclCommunicator.initAll
```

Comment:
I feel like `MeshHostWorker` class is getting way too big if we include all these functions. Can we move all the `xla_nccl_*` functions to one class and all `cupy_nccl_*` functions to another class, and then choose one class here?

Reply:
Agree. Besides, most code in `xla_nccl_XXX` can be shared with `cupy_nccl_XXX`. Maybe we can abstract each step like:
```python
def nccl_send_tile(self, buf, offset, *args):
    tensor_shape = buf.shape
    start_pos, end_pos = infer_start_pos_and_n_elements(tensor_shape, offset)
    buf = self.preprocess_buffer(buf)    // xla_to_cupy here
    self.send_multigpu(buf, start_pos, end_pos, *args)
```
The same applies for recv

Reply:
cupy nccl will be deleted soon. Could I keep it for a while? After this PR is merged for one week and keeps running smoothly, I will delete cupy nccl related code.